pkg/workloads/console/runner/runner.go

```diff
@@ -0,0 +1,162 @@
+	// Later on we may need to implement non-interactive consoles, for processes which
+	// expect a TTY to not be present?
+	// However with these types of consoles it will not be possible to send input to them
+	// when reattaching, e.g. attempting to send a SIGINT to cancel the running process.
+	// Interactive bool
+}
+
+// New builds a runner, with a given clientset
+func New(clientset versioned.Interface) *Runner {
+	return &Runner{
+		clientset: clientset,
+	}
+}
+
+// Create builds a console according to the supplied options and submits it to the API
+func (c *Runner) Create(namespace string, template workloadsv1alpha1.ConsoleTemplate, opts Options) (*workloadsv1alpha1.Console, error) {
+	csl := &workloadsv1alpha1.Console{
+		ObjectMeta: metav1.ObjectMeta{
+			// Let Kubernetes generate a unique name
+			GenerateName: template.Name + "-",
+
+			// TODO: Consider which labels and annotations we might want to set on the console
+			// itself.
+			// Labels:
+			// Annotations:
+		},
+		Spec: workloadsv1alpha1.ConsoleSpec{
+			ConsoleTemplateRef: corev1.LocalObjectReference{Name: template.Name},
+			// If the flag is not provided then the value will default to 0. The controller
+			// should detect this and apply the default timeout that is defined in the template.
+			TimeoutSeconds: opts.Timeout,
+			// TODO: This field is currently missing from the console spec
+			// Command:            c.Cmd,
+		},
+	}
+
+	client := c.clientset.WorkloadsV1alpha1().Consoles(namespace)
+	return client.Create(csl)
+}
+
+// FindTemplateBySelector will search for a template matching the given label
+// selector and return errors if none or multiple are found (when the selector
+// is too broad)
+func (c *Runner) FindTemplateBySelector(namespace string, labelSelector string) (*workloadsv1alpha1.ConsoleTemplate, error) {
+	client := c.clientset.WorkloadsV1alpha1().ConsoleTemplates(namespace)
+
+	templates, err := client.List(
+		metav1.ListOptions{
+			LabelSelector: labelSelector,
+		},
+	)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to list consoles templates")
+	}
+
+	if len(templates.Items) != 1 {
+		identifiers := []string{}
+		for _, item := range templates.Items {
+			identifiers = append(identifiers, item.Namespace+"/"+item.Name)
+		}
+
+		return nil, errors.Errorf(
+			"expected to discover 1 console template, but actually found: %s",
+			identifiers,
+		)
+	}
+
+	template := templates.Items[0]
+
+	return &template, nil
+}
+
+// WaitUntilReady will block until the console reaches a phase that indicates
+// that it's ready to be attached to, or has failed.
+func (c *Runner) WaitUntilReady(ctx context.Context, createdCsl workloadsv1alpha1.Console) error {
+	isRunning := func(csl *workloadsv1alpha1.Console) bool {
+		return csl != nil && csl.Status.Phase == workloadsv1alpha1.ConsoleRunning
+	}
+	isStopped := func(csl *workloadsv1alpha1.Console) bool {
+		return csl != nil && csl.Status.Phase == workloadsv1alpha1.ConsoleStopped
+	}
```

Comment:
There are helpers in https://github.com/gocardless/theatre/pull/31 that will help with this

Reply:
Yep that'll help. Unfortunately we may still need the nil checks though